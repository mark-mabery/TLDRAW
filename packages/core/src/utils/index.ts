import { Utils } from './utils'
export { Utils } from './utils'
export { Svg } from './svg'

export default Utils
