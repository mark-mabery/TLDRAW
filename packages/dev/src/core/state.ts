import type {
  TLBinding,
  TLPage,
  TLPageState,
  TLPointerEventHandler,
  TLShapeChangeHandler,
} from '@tldraw/core'
import type { RectangleShape } from './box'
import type { LabelShape } from './label'
import { StateManager } from 'rko'

type Shapes = RectangleShape | LabelShape

interface State {
  page: TLPage<Shapes, TLBinding>
  pageState: TLPageState
  meta: {
    isDarkMode: boolean
  }
}

class AppState extends StateManager<State> {
  /* ----------------------- API ---------------------- */

  selectShape(shapeId: string) {
    this.patchState({
      pageState: {
        selectedIds: [shapeId],
      },
    })
  }

  deselect() {
    this.patchState({
      pageState: {
        selectedIds: [],
        editingId: undefined,
      },
    })
  }

  startEditingShape(shapeId: string) {
    this.patchState({
      pageState: {
        selectedIds: [shapeId],
        editingId: shapeId,
      },
    })
  }

  changeShapeText = (id: string, text: string) => {
    this.patchState({
      page: {
        shapes: {
          [id]: { text },
        },
      },
    })
  }

  /* --------------------- Events --------------------- */

  onPointCanvas: TLPointerEventHandler = (info) => {
    this.deselect()
  }

  onPointShape: TLPointerEventHandler = (info) => {
    this.selectShape(info.target)
  }

  onDoubleClickShape: TLPointerEventHandler = (info) => {
    this.startEditingShape(info.target)
  }

  onDoubleClickBounds: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerDown: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerUp: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerMove: TLPointerEventHandler = (info) => {
    // Todo
  }

  onShapeChange: TLShapeChangeHandler<Shapes> = (shape) => {
    if (shape.type === 'rectangle' && shape.size) {
      this.patchState({
        page: {
          shapes: {
            [shape.id]: { ...shape, size: [...shape.size] },
          },
        },
      })
    }
  }
}

export const appState = new AppState({
  page: {
    id: 'page1',
    shapes: {
      rect1: {
        id: 'rect1',
        parentId: 'page1',
        name: 'Rectangle',
        childIndex: 1,
        type: 'rectangle',
        point: [0, 0],
        rotation: 0,
        size: [100, 100],
        text: 'Hello world!',
      },
      label1: {
        id: 'label1',
        parentId: 'page1',
        name: 'Label',
        childIndex: 2,
        type: 'label',
        point: [-200, -200],
        rotation: 0,
        text: 'My shape is stateful, I should still render while off screen!',
      },
      label2: {
        id: 'label2',
        parentId: 'page1',
        name: 'Label',
        childIndex: 2,
        type: 'label',
        point: [200, 200],
        rotation: 0,
        text: 'Hello world!',
      },
    },
    bindings: {},
  },
  pageState: {
    id: 'page1',
    selectedIds: [],
    camera: {
      point: [0, 0],
      zoom: 1,
    },
  },
  meta: {
    isDarkMode: false,
  },
})
