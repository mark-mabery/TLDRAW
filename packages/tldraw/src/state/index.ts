export * from './tlstate'
export * from './state-types'
