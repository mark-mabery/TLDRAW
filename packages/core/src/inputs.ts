import type React from 'react'
import type { TLKeyboardInfo, TLPointerInfo } from './types'
import { Vec, Utils } from './utils'

const DOUBLE_CLICK_DURATION = 250

export class Inputs {
  pointer?: TLPointerInfo<string>
  keyboard?: TLKeyboardInfo
  keys: Record<string, boolean> = {}
  isPinching = false

  offset = [0, 0]
  size = [10, 10]

  pointerUpTime = 0

  activePointer?: number

  pointerIsValid(e: TouchEvent | React.TouchEvent | PointerEvent | React.PointerEvent) {
    if ('pointerId' in e) {
      if (this.activePointer && this.activePointer !== e.pointerId) return false
    }

    if ('touches' in e) {
      const touch = e.changedTouches[0]
      if (this.activePointer && this.activePointer !== touch.identifier) return false
    }

    return true
  }

  touchStart<T extends string>(e: TouchEvent | React.TouchEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const touch = e.changedTouches[0]

    this.activePointer = touch.identifier

    const info: TLPointerInfo<T> = {
      target,
      pointerId: touch.identifier,
      origin: Inputs.getPoint(touch),
      delta: [0, 0],
      point: Inputs.getPoint(touch),
      pressure: Inputs.getPressure(touch),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  touchEnd<T extends string>(e: TouchEvent | React.TouchEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const touch = e.changedTouches[0]

    const info: TLPointerInfo<T> = {
      target,
      pointerId: touch.identifier,
      origin: Inputs.getPoint(touch),
      delta: [0, 0],
      point: Inputs.getPoint(touch),
      pressure: Inputs.getPressure(touch),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    this.activePointer = undefined

    return info
  }

  touchMove<T extends string>(e: TouchEvent | React.TouchEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const touch = e.changedTouches[0]

    const prev = this.pointer

    const point = Inputs.getPoint(touch)

    const delta = prev?.point ? Vec.sub(point, prev.point) : [0, 0]

    const info: TLPointerInfo<T> = {
      origin: point,
      ...prev,
      target,
      pointerId: touch.identifier,
      point,
      delta,
      pressure: Inputs.getPressure(touch),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerDown<T extends string>(e: PointerEvent | React.PointerEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const point = Inputs.getPoint(e, this.offset)

    this.activePointer = e.pointerId

    const info: TLPointerInfo<T> = {
      target,
      pointerId: e.pointerId,
      origin: point,
      point: point,
      delta: [0, 0],
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerEnter<T extends string>(
    e: PointerEvent | React.PointerEvent,
    target: T
  ): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const point = Inputs.getPoint(e, this.offset)

    const info: TLPointerInfo<T> = {
      target,
      pointerId: e.pointerId,
      origin: point,
      delta: [0, 0],
      point: point,
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerMove<T extends string>(e: PointerEvent | React.PointerEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const prev = this.pointer

    const point = Inputs.getPoint(e, this.offset)

    const delta = prev?.point ? Vec.sub(point, prev.point) : [0, 0]

    const info: TLPointerInfo<T> = {
      origin: point,
      ...prev,
      target,
      pointerId: e.pointerId,
      point,
      delta,
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pointerUp<T extends string>(e: PointerEvent | React.PointerEvent, target: T): TLPointerInfo<T> {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const prev = this.pointer

    const point = Inputs.getPoint(e, this.offset)

    const delta = prev?.point ? Vec.sub(point, prev.point) : [0, 0]

    this.activePointer = undefined

    const info: TLPointerInfo<T> = {
      origin: point,
      ...prev,
      target,
      pointerId: e.pointerId,
      point,
      delta,
      pressure: Inputs.getPressure(e),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    this.pointerUpTime = Date.now()

    return info
  }

  panStart = (e: WheelEvent): TLPointerInfo<'wheel'> => {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const info: TLPointerInfo<'wheel'> = {
      target: 'wheel',
      pointerId: this.pointer?.pointerId || 0,
      origin: this.pointer?.origin || [0, 0],
      delta: [0, 0],
      pressure: 0.5,
      point: Inputs.getPoint(e, this.offset),
      shiftKey,
      ctrlKey,
      metaKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  pan = (delta: number[], e: WheelEvent): TLPointerInfo<'wheel'> => {
    if (!this.pointer || this.pointer.target !== 'wheel') {
      return this.panStart(e)
    }

    const { shiftKey, ctrlKey, metaKey, altKey } = e

    const prev = this.pointer

    const point = Inputs.getPoint(e, this.offset)

    const info: TLPointerInfo<'wheel'> = {
      ...prev,
      target: 'wheel',
      delta,
      point,
      shiftKey,
      ctrlKey,
      metaKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  isDoubleClick() {
    if (!this.pointer) return false

    const { origin, point } = this.pointer

    return Date.now() - this.pointerUpTime < DOUBLE_CLICK_DURATION && Vec.dist(origin, point) < 4
  }

  clear() {
    this.pointer = undefined
  }

  resetDoubleClick() {
    this.pointerUpTime = 0
  }

  keydown = (e: KeyboardEvent | React.KeyboardEvent): TLKeyboardInfo => {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    this.keys[e.key] = true

    return {
      point: this.pointer?.point || [0, 0],
      origin: this.pointer?.origin || [0, 0],
      key: e.key,
      keys: Object.keys(this.keys),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }
  }

  keyup = (e: KeyboardEvent | React.KeyboardEvent): TLKeyboardInfo => {
    const { shiftKey, ctrlKey, metaKey, altKey } = e

    delete this.keys[e.key]

    return {
      point: this.pointer?.point || [0, 0],
      origin: this.pointer?.origin || [0, 0],
      key: e.key,
      keys: Object.keys(this.keys),
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }
  }

  pinch(point: number[], origin: number[]) {
    const { shiftKey, ctrlKey, metaKey, altKey } = this.keys

    const delta = Vec.sub(origin, point)

    const info: TLPointerInfo<'pinch'> = {
      pointerId: 0,
      target: 'pinch',
      origin,
      delta: delta,
      point: Vec.sub(Vec.round(point), this.offset),
      pressure: 0.5,
      shiftKey,
      ctrlKey,
      metaKey: Utils.isDarwin() ? metaKey : ctrlKey,
      altKey,
    }

    this.pointer = info

    return info
  }

  reset() {
    this.pointerUpTime = 0
    this.pointer = undefined
    this.keyboard = undefined
    this.activePointer = undefined
    this.keys = {}
  }

  static getPoint(
    e: PointerEvent | React.PointerEvent | Touch | React.Touch | WheelEvent,
    offset = [0, 0]
  ): number[] {
    return [
      Number(e.clientX.toPrecision(5)) - offset[0],
      Number(e.clientY.toPrecision(5)) - offset[1],
    ]
  }

  static getPressure(e: PointerEvent | React.PointerEvent | Touch | React.Touch | WheelEvent) {
    return 'pressure' in e ? Number(e.pressure.toPrecision(5)) || 0.5 : 0.5
  }

  static commandKey(): string {
    return Utils.isDarwin() ? 'âŒ˜' : 'Ctrl'
  }
}

export const inputs = new Inputs()
