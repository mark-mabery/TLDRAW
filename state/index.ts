import state, { useSelector } from "./state"

export default state

export { useSelector }
