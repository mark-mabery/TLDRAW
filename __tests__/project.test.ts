import * as json from './__mocks__/document.json'
import state from 'state'
import { point } from './test-utils'
import inputs from 'state/inputs'
import { getSelectedIds, setToArray } from 'utils/utils'

const rectangleId = '1f6c251c-e12e-40b4-8dd2-c1847d80b72f'
const arrowId = '5ca167d7-54de-47c9-aa8f-86affa25e44d'

describe('project', () => {
  it('mounts the state', () => {
    state.enableLog(true)

    state
      .send('MOUNTED')
      .send('LOADED_FROM_FILE', { json: JSON.stringify(json) })
  })

  it('selects and deselects a shape', () => {
    expect(setToArray(getSelectedIds(state.data))).toStrictEqual([])

    state
      .send('POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
      .send('STOPPED_POINTING', inputs.pointerUp(point()))

    expect(setToArray(getSelectedIds(state.data))).toStrictEqual([rectangleId])

    state
      .send('POINTED_CANVAS', inputs.pointerDown(point(), 'canvas'))
      .send('STOPPED_POINTING', inputs.pointerUp(point()))

    expect(setToArray(getSelectedIds(state.data))).toStrictEqual([])
  })

  it('selects multiple shapes', () => {
    expect(setToArray(getSelectedIds(state.data))).toStrictEqual([])

    state
      .send('POINTED_SHAPE', inputs.pointerDown(point(), rectangleId))
      .send('STOPPED_POINTING', inputs.pointerUp(point()))

    expect(setToArray(getSelectedIds(state.data))).toStrictEqual([rectangleId])

    state.send(
      'POINTED_SHAPE',
      inputs.pointerDown(point({ shiftKey: true }), arrowId)
    )

    // state.send('STOPPED_POINTING', inputs.pointerUp(point()))

    expect(setToArray(getSelectedIds(state.data))).toStrictEqual([
      rectangleId,
      arrowId,
    ])
  })
})
