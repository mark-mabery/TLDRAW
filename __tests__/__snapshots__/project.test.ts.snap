// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project loads file from json: data after mount from file 1`] = `
Object {
  "code": Object {
    "file0": Object {
      "code": "
const draw = new Draw({
  points: [
    ...Utils.getPointsBetween([0, 0], [20, 50]),
    ...Utils.getPointsBetween([20, 50], [100, 20], 3),
    ...Utils.getPointsBetween([100, 20], [100, 100], 10),
    [100, 100],
  ],
})

const rectangle = new Rectangle({
  point: [200, 0],
  style: {
    color: ColorStyle.Blue,
  },
})

const ellipse = new Ellipse({
  point: [400, 0],
})

const arrow = new Arrow({
  start: [600, 0],
  end: [700, 100],
})

const radius = 1000
const count = 100
const center = [350, 50]

for (let i = 0; i < count; i++) {
  const point = Vec.rotWith(
    Vec.add(center, [radius, 0]),
    center,
    (Math.PI * 2 * i) / count
  )

  const dot = new Dot({
    point,
  })
}
        ",
      "id": "file0",
      "name": "index.ts",
    },
  },
  "id": "home",
  "name": "My Document",
  "pages": Object {
    "page1": Object {
      "childIndex": 0,
      "id": "page1",
      "name": "Page 1",
      "shapes": Object {
        "13448777-d8f5-46cd-8a70-a4259211902e": Object {
          "childIndex": 4,
          "id": "13448777-d8f5-46cd-8a70-a4259211902e",
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            511.7,
            404.19,
          ],
          "radius": 2,
          "rotation": 0,
          "size": Array [
            181.08999999999992,
            150.40999999999997,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "rectangle",
        },
        "2d842ace-ebc5-4e83-acdf-de29352e5e62": Object {
          "bend": 0,
          "childIndex": 10,
          "decorations": Object {
            "end": "Arrow",
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                1.2250000000000227,
                92.60000000000002,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                2.4500000000000455,
                185.20000000000005,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                0,
                0,
              ],
            },
          },
          "id": "2d842ace-ebc5-4e83-acdf-de29352e5e62",
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            616.9,
            1124.3,
          ],
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "arrow",
        },
        "38e9e750-16c2-4476-93ab-21aeb5f8858f": Object {
          "childIndex": 12,
          "id": "38e9e750-16c2-4476-93ab-21aeb5f8858f",
          "name": "Text",
          "parentId": "page1",
          "point": Array [
            207.16,
            1422.4,
          ],
          "rotation": 0,
          "scale": 1,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "text": "Hello",
          "type": "text",
        },
        "3c688979-b190-4270-915b-7d8dd22a2bb7": Object {
          "childIndex": 14,
          "id": "3c688979-b190-4270-915b-7d8dd22a2bb7",
          "name": "Text",
          "parentId": "page1",
          "point": Array [
            564.06,
            1558.1,
          ],
          "rotation": 0,
          "scale": 1,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "text": "Hello",
          "type": "text",
        },
        "51641de1-9787-41b8-afcc-2c85fd1b24c7": Object {
          "childIndex": 7,
          "id": "51641de1-9787-41b8-afcc-2c85fd1b24c7",
          "name": "Ellipse",
          "parentId": "page1",
          "point": Array [
            517.18,
            783.54,
          ],
          "radiusX": 102.99999999999997,
          "radiusY": 102.99999999999994,
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "ellipse",
        },
        "5ba998df-c036-447a-9b88-d96c71394f52": Object {
          "childIndex": 13,
          "id": "5ba998df-c036-447a-9b88-d96c71394f52",
          "name": "Text",
          "parentId": "page1",
          "point": Array [
            389.57,
            1496.5,
          ],
          "rotation": 0,
          "scale": 1,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "text": "Hello",
          "type": "text",
        },
        "75010635-8dfb-48ea-9250-719e50e58f02": Object {
          "childIndex": 5,
          "id": "75010635-8dfb-48ea-9250-719e50e58f02",
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            384.09,
            378.45,
          ],
          "radius": 2,
          "rotation": 0,
          "size": Array [
            95.20999999999992,
            91.1799999999999,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "rectangle",
        },
        "b8e4e2c5-c662-4587-bf80-b9820ad8ad7f": Object {
          "bend": 0,
          "childIndex": 11,
          "decorations": Object {
            "end": "Arrow",
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                0.9250000000000114,
                47.85000000000002,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                1.8500000000000227,
                95.70000000000005,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                0,
                0,
              ],
            },
          },
          "id": "b8e4e2c5-c662-4587-bf80-b9820ad8ad7f",
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            425.18,
            1143.2,
          ],
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "arrow",
        },
        "c892d665-3311-4e25-a0bf-c4632d777f7e": Object {
          "childIndex": 6,
          "id": "c892d665-3311-4e25-a0bf-c4632d777f7e",
          "name": "Ellipse",
          "parentId": "page1",
          "point": Array [
            162.45,
            679.23,
          ],
          "radiusX": 102.99999999999997,
          "radiusY": 102.99999999999994,
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "ellipse",
        },
        "e08c415e-3db3-4d3b-878e-28ce693ec1b0": Object {
          "childIndex": 8,
          "id": "e08c415e-3db3-4d3b-878e-28ce693ec1b0",
          "name": "Ellipse",
          "parentId": "page1",
          "point": Array [
            398.99,
            810.79,
          ],
          "radiusX": 45.484999999999985,
          "radiusY": 45.48499999999996,
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "ellipse",
        },
        "e43559cb-6f41-4ae4-9c49-158ed1ad2f72": Object {
          "childIndex": 3,
          "id": "e43559cb-6f41-4ae4-9c49-158ed1ad2f72",
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            171.47,
            288.63,
          ],
          "radius": 2,
          "rotation": 0,
          "size": Array [
            176.22,
            192.26,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "rectangle",
        },
        "fee77127-e779-4576-882b-b1bf7c7e132f": Object {
          "bend": 0,
          "childIndex": 9,
          "decorations": Object {
            "end": "Arrow",
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                0.045000000000001705,
                104,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                0.09000000000000341,
                208,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                0,
                0,
              ],
            },
          },
          "id": "fee77127-e779-4576-882b-b1bf7c7e132f",
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            252.85,
            1057.5,
          ],
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "arrow",
        },
      },
      "type": "page",
    },
  },
}
`;

exports[`restoring project remounts the state after mutating the current state: data after re-mount from file 1`] = `
Object {
  "code": Object {
    "file0": Object {
      "code": "
const draw = new Draw({
  points: [
    ...Utils.getPointsBetween([0, 0], [20, 50]),
    ...Utils.getPointsBetween([20, 50], [100, 20], 3),
    ...Utils.getPointsBetween([100, 20], [100, 100], 10),
    [100, 100],
  ],
})

const rectangle = new Rectangle({
  point: [200, 0],
  style: {
    color: ColorStyle.Blue,
  },
})

const ellipse = new Ellipse({
  point: [400, 0],
})

const arrow = new Arrow({
  start: [600, 0],
  end: [700, 100],
})

const radius = 1000
const count = 100
const center = [350, 50]

for (let i = 0; i < count; i++) {
  const point = Vec.rotWith(
    Vec.add(center, [radius, 0]),
    center,
    (Math.PI * 2 * i) / count
  )

  const dot = new Dot({
    point,
  })
}
        ",
      "id": "file0",
      "name": "index.ts",
    },
  },
  "id": "home",
  "name": "My Document",
  "pages": Object {
    "page1": Object {
      "childIndex": 0,
      "id": "page1",
      "name": "Page 1",
      "shapes": Object {
        "13448777-d8f5-46cd-8a70-a4259211902e": Object {
          "childIndex": 4,
          "id": "13448777-d8f5-46cd-8a70-a4259211902e",
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            511.7,
            404.19,
          ],
          "radius": 2,
          "rotation": 0,
          "size": Array [
            181.08999999999992,
            150.40999999999997,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "rectangle",
        },
        "2d842ace-ebc5-4e83-acdf-de29352e5e62": Object {
          "bend": 0,
          "childIndex": 10,
          "decorations": Object {
            "end": "Arrow",
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                1.2250000000000227,
                92.60000000000002,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                2.4500000000000455,
                185.20000000000005,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                0,
                0,
              ],
            },
          },
          "id": "2d842ace-ebc5-4e83-acdf-de29352e5e62",
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            616.9,
            1124.3,
          ],
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "arrow",
        },
        "38e9e750-16c2-4476-93ab-21aeb5f8858f": Object {
          "childIndex": 12,
          "id": "38e9e750-16c2-4476-93ab-21aeb5f8858f",
          "name": "Text",
          "parentId": "page1",
          "point": Array [
            207.16,
            1422.4,
          ],
          "rotation": 0,
          "scale": 1,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "text": "Hello",
          "type": "text",
        },
        "3c688979-b190-4270-915b-7d8dd22a2bb7": Object {
          "childIndex": 14,
          "id": "3c688979-b190-4270-915b-7d8dd22a2bb7",
          "name": "Text",
          "parentId": "page1",
          "point": Array [
            564.06,
            1558.1,
          ],
          "rotation": 0,
          "scale": 1,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "text": "Hello",
          "type": "text",
        },
        "51641de1-9787-41b8-afcc-2c85fd1b24c7": Object {
          "childIndex": 7,
          "id": "51641de1-9787-41b8-afcc-2c85fd1b24c7",
          "name": "Ellipse",
          "parentId": "page1",
          "point": Array [
            517.18,
            783.54,
          ],
          "radiusX": 102.99999999999997,
          "radiusY": 102.99999999999994,
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "ellipse",
        },
        "5ba998df-c036-447a-9b88-d96c71394f52": Object {
          "childIndex": 13,
          "id": "5ba998df-c036-447a-9b88-d96c71394f52",
          "name": "Text",
          "parentId": "page1",
          "point": Array [
            389.57,
            1496.5,
          ],
          "rotation": 0,
          "scale": 1,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "text": "Hello",
          "type": "text",
        },
        "75010635-8dfb-48ea-9250-719e50e58f02": Object {
          "childIndex": 5,
          "id": "75010635-8dfb-48ea-9250-719e50e58f02",
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            384.09,
            378.45,
          ],
          "radius": 2,
          "rotation": 0,
          "size": Array [
            95.20999999999992,
            91.1799999999999,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "rectangle",
        },
        "b8e4e2c5-c662-4587-bf80-b9820ad8ad7f": Object {
          "bend": 0,
          "childIndex": 11,
          "decorations": Object {
            "end": "Arrow",
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                0.9250000000000114,
                47.85000000000002,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                1.8500000000000227,
                95.70000000000005,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                0,
                0,
              ],
            },
          },
          "id": "b8e4e2c5-c662-4587-bf80-b9820ad8ad7f",
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            425.18,
            1143.2,
          ],
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "arrow",
        },
        "c892d665-3311-4e25-a0bf-c4632d777f7e": Object {
          "childIndex": 6,
          "id": "c892d665-3311-4e25-a0bf-c4632d777f7e",
          "name": "Ellipse",
          "parentId": "page1",
          "point": Array [
            162.45,
            679.23,
          ],
          "radiusX": 102.99999999999997,
          "radiusY": 102.99999999999994,
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "ellipse",
        },
        "e08c415e-3db3-4d3b-878e-28ce693ec1b0": Object {
          "childIndex": 8,
          "id": "e08c415e-3db3-4d3b-878e-28ce693ec1b0",
          "name": "Ellipse",
          "parentId": "page1",
          "point": Array [
            398.99,
            810.79,
          ],
          "radiusX": 45.484999999999985,
          "radiusY": 45.48499999999996,
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "ellipse",
        },
        "e43559cb-6f41-4ae4-9c49-158ed1ad2f72": Object {
          "childIndex": 3,
          "id": "e43559cb-6f41-4ae4-9c49-158ed1ad2f72",
          "name": "Rectangle",
          "parentId": "page1",
          "point": Array [
            171.47,
            288.63,
          ],
          "radius": 2,
          "rotation": 0,
          "size": Array [
            176.22,
            192.26,
          ],
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "rectangle",
        },
        "fee77127-e779-4576-882b-b1bf7c7e132f": Object {
          "bend": 0,
          "childIndex": 9,
          "decorations": Object {
            "end": "Arrow",
            "middle": null,
            "start": null,
          },
          "handles": Object {
            "bend": Object {
              "id": "bend",
              "index": 2,
              "point": Array [
                0.045000000000001705,
                104,
              ],
            },
            "end": Object {
              "id": "end",
              "index": 1,
              "point": Array [
                0.09000000000000341,
                208,
              ],
            },
            "start": Object {
              "id": "start",
              "index": 0,
              "point": Array [
                0,
                0,
              ],
            },
          },
          "id": "fee77127-e779-4576-882b-b1bf7c7e132f",
          "name": "Arrow",
          "parentId": "page1",
          "point": Array [
            252.85,
            1057.5,
          ],
          "rotation": 0,
          "style": Object {
            "color": "Black",
            "dash": "Draw",
            "isFilled": false,
            "size": "Medium",
          },
          "type": "arrow",
        },
      },
      "type": "page",
    },
  },
}
`;
